multiuser/api/che-multiuser-api-remote-subscription/src/main/java/org/eclipse/che/multiuser/api/subscription/DistributedRemoteSubscriptionStorage.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.multiuser.api.subscription;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.notification.RemoteSubscriptionContext;
+import org.eclipse.che.api.core.notification.RemoteSubscriptionStorage;
+import org.jgroups.JChannel;
+import org.jgroups.blocks.ReplicatedHashMap;
+import org.jgroups.blocks.locking.LockService;
+import org.slf4j.Logger;
+
+/**
+ * Replicated map-based implementation of {@link RemoteSubscriptionStorage}
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class DistributedRemoteSubscriptionStorage implements RemoteSubscriptionStorage {
+
+  private static final Logger LOG = getLogger(DistributedRemoteSubscriptionStorage.class);
+
+  private static final String CHANNEL_NAME = "RemoteSubscriptionChannel";
+
+  private ReplicatedHashMap<String, Set<RemoteSubscriptionContext>> subscriptions;
+
+  private final LockService lockService;
+
+  @Inject
+  public DistributedRemoteSubscriptionStorage(@Named("jgroups.config.file") String confFile)
+      throws Exception {
+    try {
+      JChannel channel = new JChannel(confFile);
+      this.lockService = new LockService(channel);
+      channel.connect(CHANNEL_NAME);
+      subscriptions = new ReplicatedHashMap<>(channel);
```

Comment:
looks like we are trying to achieve the consistent state of subscribers in each period of time, it looks like updates should be performed across the cluster synchronously. to achieve that use blocking_updates property, if we won't need that we can manage local concurrent operations using read/write locks.

Reply:
Makes sense